src/promptflow/promptflow/_sdk/_telemetry/logging_handler.py

```diff
@@ -73,71 +53,127 @@ def get_scrubbed_cloud_role():
             return
 
-    def log_record_to_envelope(self, record):
+    @staticmethod
+    def _get_attributes(record: logging.LogRecord) -> Attributes:
+        """Get the attributes from the log record.
+        Since the value in Attributes cannot be dict,
+        otherwise the dict value will be filtered out, so it is necessary to parse dic in this step.
+        """
         from promptflow._utils.utils import is_in_ci_pipeline
 
-        # skip logging if telemetry is disabled
-
-        if not self._is_telemetry_enabled:
-            return
         custom_dimensions = {
             "level": record.levelname,
             # add to distinguish if the log is from ci pipeline
             "from_ci": is_in_ci_pipeline(),
         }
-        custom_dimensions.update(self._custom_dimensions)
-        if hasattr(record, "custom_dimensions") and isinstance(record.custom_dimensions, dict):
-            record.custom_dimensions.update(custom_dimensions)
-        else:
-            record.custom_dimensions = custom_dimensions
 
-        envelope = super().log_record_to_envelope(record=record)
-        # scrub data before sending to appinsights
-        role = get_scrubbed_cloud_role()
-        envelope.tags["ai.cloud.role"] = role
-        envelope.tags.pop("ai.cloud.roleInstance", None)
-        envelope.tags.pop("ai.device.id", None)
-        return envelope
+        attributes = LoggingHandler._get_attributes(record)
+        return {
+            # Mark the data as CustomEvents and record the data to CustomEvents.
+            _APPLICATION_INSIGHTS_EVENT_MARKER_ATTRIBUTE: True,
+            **custom_dimensions,
+            **attributes.get("custom_dimensions", {}),
+        }
 
     @classmethod
     def disable_telemetry_logger(cls):
-        """Disable AzureEventHandler's logging to avoid warning affect user experience"""
-        from opencensus.ext.azure.common.processor import logger as processor_logger
-        from opencensus.ext.azure.common.storage import logger as storage_logger
-        from opencensus.ext.azure.common.transport import logger as transport_logger
+        """Disable opentelemetry's logging to avoid warning affect user experience"""
+
+        from azure.monitor.opentelemetry.exporter.export.logs._exporter import _logger as exporter_logger
+        from opentelemetry._logs._internal import _logger as provider_logger
+        from opentelemetry.sdk._logs._internal import _logger as handler_logger
+        from opentelemetry.sdk._logs._internal.export import _logger as processor_logger
 
+        handler_logger.setLevel(logging.CRITICAL)
         processor_logger.setLevel(logging.CRITICAL)
-        transport_logger.setLevel(logging.CRITICAL)
-        storage_logger.setLevel(logging.CRITICAL)
+        provider_logger.setLevel(logging.CRITICAL)
+        exporter_logger.setLevel(logging.CRITICAL)
+
+
+@cache_result_with_specify_result_type(result_type=PromptFlowSDKLogHandler)
```

Comment:
why do we need to cache the result now? Do we find any perf issue?

Reply:
Just for the sake of optimizing the code, if we need to ensure that this handler is the same in our code and not duplicated, we can cache the result of this function to ensure that the handler obtained everywhere is the same, and the code is also more concise.

![image](https://github.com/microsoft/promptflow/assets/23182548/f967e102-1d48-447e-8792-b6bbf524f1d3)

Reply:
Are we trying to achieve a singleton behavior here? Maybe just store the value in a static field?

Reply:
Do you mean to use a static variable to save the result of this **get_appinsights_log_handler**? But I think the current way of decorating functions with decorators and caching results is more elegant? And this caching function is also more universal.

Reply:
Based on offline discussions, use **functions. lru_cache**